pyro/distributions/transforms/naf.py

```diff
@@ -13,12 +13,123 @@
 eps = 1e-8
 
 
+class ELUMixin(object):
+    @staticmethod
+    def f(x):
+        """
+        Implements the nonlinearity of NAF, in this case ELU
+        """
+        return F.elu(x)
+
+    @staticmethod
+    def f_inv(x):
+        """
+        Implements the inverse of ELU
+        """
+        return torch.max(x, torch.zeros_like(x)) + torch.min(torch.log1p(x + eps), torch.zeros_like(x))
+
+    @staticmethod
+    def log_df_dx(x):
+        """
+        Implements the log derivative of NAF nonlinearity
+        """
+        return -F.relu(-x)
+
+    @staticmethod
+    def log_df_inv_dx(x):
+        """
+        Implements the log derivative of inverse NAF nonlinearity
+        """
+        return F.relu(-torch.log1p(x + eps))
+
+
+class LeakyReLUMixin(object):
+    @staticmethod
+    def f(x):
+        """
+        Implements the nonlinearity of NAF, in this case leaky ReLU
+        """
+        return F.leaky_relu(x)
+
+    @staticmethod
+    def f_inv(x):
+        """
+        Implements the inverse of leaky ReLU
+        """
+        # slope for negative part is inverse of slope for positive part in f(x)
+        return F.leaky_relu(x, negative_slope=100.0)
+
+    @staticmethod
+    def log_df_dx(x):
+        """
+        Implements the log derivative of NAF nonlinearity
+        """
+        return torch.where(x >= 0., torch.zeros_like(x), torch.ones_like(x) * math.log(0.01))
+
+    @staticmethod
+    def log_df_inv_dx(x):
+        """
+        Implements the log derivative of inverse NAF nonlinearity
+        """
+        return torch.where(x >= 0., torch.zeros_like(x), torch.ones_like(x) * math.log(100.0))
+
+
+def safe_log(x):
```

Comment:
Just curious - in which cases do you find `safe_log` to be useful?

Reply:
I think the intention for for very small `x`